File: solutions/091.Decode_Ways/AC_dp_n.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_dp_n.py
# Create Date: 2015-02-26 12:44:05
# Usage:       AC_dp_n.py 
# Descripton:  


class Solution:
    # @param s, a string
    # @return an integer
    def numDecodings(self, s):
        if not s:
            return 0
        if len(s) == 1:
            return 1 if s[0] != '0' else 0

        dp = [1, 1 if s[0] != '0' else 0]
        for i in range(2, len(s) + 1):
            rec = 0
            if s[i - 1] != '0':
                rec += dp[(i-1)&1]
            if s[i - 2 : i] <= '26' and s[i - 2] != '0':
                rec += dp[i&1]
            dp[i&1] = rec
        return dp[len(s)&1]

# debug
s = Solution()
print s.numDecodings("12")

File: solutions/055.Jump_Game/AC_dp_n.cpp
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_dp_n.cpp
*  Create Date: 2015-01-30 09:46:40
*  Descripton:  f[i] means the max steps can go at i
*               f[i] = max(f[i - 1], A[i - 1]) - 1
*/

#include <bits/stdc++.h>

using namespace std;
const int N = 0;

class Solution {
public:
    bool canJump(int A[], int n) {
        // vector<int> f(n, 0);
        int f[2];        // compress to O(1)
        f[0] = 0;
        for (int i = 1; i < n; ++i) {
            f[i&1] = max(f[(i-1)&1], A[i - 1]) - 1;
            if (f[i&1] < 0)
                return false;
        }
        return f[(n-1)&1] >= 0;
    }
};

int main() {
    int n, A[1000];
    Solution s;
    while (cin >> n) {
        for (int i = 0; i < n; i++)
            cin >> A[i];
        cout << s.canJump(A, n) << endl;
    }
    return 0;
}

